snuba/cli/bootstrap.py

```diff
@@ -106,11 +108,33 @@ def bootstrap(
     # tables or distributed tables, etc.
 
     # Create the tables for every dataset.
+    existing_tables = {row[0] for row in conn.execute("show tables")}
     for name in DATASET_NAMES:
         dataset = get_dataset(name)
 
         logger.debug("Creating tables for dataset %s", name)
+        run_migrations = False
         for statement in dataset.get_dataset_schemas().get_create_statements():
-            logger.debug("Executing:\n%s", statement)
-            ClickhousePool().execute(statement)
+            if statement.table_name not in existing_tables:
+                # This is a hack to deal with updates to Materialized views.
+                # It seems that ClickHouse would parse the SELECT statement that defines a
+                # materialized view even if the view already exists and the CREATE statement
+                # includes the IF NOT EXISTS clause.
+                # When we add a column to a matview, though, we will be in a state where, by
+                # running bootstrap, ClickHouse will parse the SQL statement to try to create
+                # the view and fail because the column does not exist yet on the underlying table,
+                # since the migration on the underlying table has not ran yet.
+                # Migrations are per dataset so they can only run after the bootstrap of an
+                # entire dataset has run. So we would have bootstrap depending on migration
+                # and migration depending on bootstrap.
+                # In order to break this dependency we skip bootstrap DDL calls here if the
+                # table/view already exists, so it is always safe to run bootstrap first.
```

Comment:
If the materialized view depends on the source table having all of the correct columns, doesn't this mean that the source table actually needs to be migrated prior to the materialized view being created? Wouldn't this cause a problem if a materialized view is added to an existing dataset, but uses a column that has not yet been migrated Take this example:

- dataset is created, with only source table
- user runs bootstrap, creating source table
- column x is added to source table
- materialized view is added to dataset utilizing column x
- time goes by, user eventually pulls latest snuba
- user runs bootstrap again, materialized view is attempted to be created, but column x does not exist, so bootstrap fails

Reply:
Indeed, that would not be covered (but this case would be broken today as well). I think the only way to manage that is that we redesign (ok, we `design`) migrations and we split migrations per table instead of defining them per dataset, so we can run bootstrap+migrations for each table.

Reply:
I guess this is fine in that it fixes the immediate problem, but we should really make sure we take care of this before it bites us, especially with on-prem and single tenant in the mix now since those are out of our normal deploy process.